### PR TITLE
added get_pdf method

### PR DESCRIPTION
**Pull request by Greyoxide**
Hi this is my first ever pull request so let me know if I've done something wrong.

I've implemented a `get_pdf`method as described in [issue number 62](https://github.com/minimul/qbo_api/issues/62).  I've also added a test to the test suite which tests this new method against the invoices endpoint.  I've also tested the new method against the development mode on my production app by downloading a pdf copy of an invoice, estimate, and purchase order.

In addition to the `get_pdf` method appended to `qbo_api/api_methods.rb`, I've also adjusted the response method found in `connection.rb`.  Given that application/json is the default content type, I've added a check to determine if this case is an exception to the default content type.

### CODE REVIEW

**Comment by minimul** [comment]
I will review within the next week.

**Comment by minimul** [comment]
OK. This looks fine.

Next week, I'm going to test it more thoroughly.

In the meantime, if you could implement my comments.

Thank you.

File: lib/qbo_api/connection.rb [hunk 1]
```diff
@@ -63,9 +63,25 @@ def raw_request(method, conn:, path:, payload: nil, params: nil, headers: nil)
...
+    # def response(resp, entity: nil)
+    #   data = resp.body
+    #   entity ? entity_response(data, entity) : data
+    # rescue => e
+    #   QboApi.logger.debug { "#{LOG_TAG} response parsing error: entity=#{entity.inspect} body=#{resp.body.inspect} exception=#{e.inspect}" }
+    #   data
+    # end
+
+    def response(resp, entity: nil, headers: nil)
+      puts headers
```
Comment (minimul) [inline lib/qbo_api/connection.rb:67]: Remove.

**Review by minimul — CHANGES_REQUESTED** [review]

File: spec/api_methods_spec.rb [hunk 2]
```diff
@@ -243,4 +243,17 @@
...
   end
+
+  describe 'get pdf' do
+    let(:entity) { :invoice }
+    let(:id) { '130' }
+
+    it 'requests the pdf of an invoice' do
+      use_cassette('get_pdf/invoice') do
+        result = api.get_pdf(entity, id)
+        pdf_starts_correctly = !!(result =~ /^%PDF-\d+(\.\d+)?/)
```
[the remark this answers is not in the document]
Comment (Greyoxide) [reply spec/api_methods_spec.rb]: good call, I've updated this spec to use `.match` to check for the pdf content response

File: lib/qbo_api/connection.rb [hunk 3]
```diff
@@ -63,16 +63,17 @@ def raw_request(method, conn:, path:, payload: nil, params: nil, headers: nil)
...
       data = resp.body
-      content_type = headers['Accept'] || headers['Content-Type'] if headers
 
-      if content_type and content_type.include?('application/pdf')
-        data # return raw pdf stream
-      else
-        # headers are optional, must assume that application/json is default
-        entity ? entity_response(data, entity) : data
+      if headers
+        # headers are an optional argument
```
Comment (minimul) [inline lib/qbo_api/connection.rb:70]: We don't need the comments. Please remove them here and in the spec.

File: spec/api_methods_spec.rb [hunk 4]
```diff
@@ -243,4 +243,18 @@
...
   end
+
+  describe 'get pdf' do
+    let(:entity) { :invoice }
+    let(:id) { '130' }
+
+    it 'requests the pdf of an invoice' do
+      use_cassette('get_pdf/invoice') do
+        result = api.get_pdf(entity, id)
+        # Ensure result is not nil and is a pdf stream
```
Comment (minimul) [inline spec/api_methods_spec.rb:254]: Remove.

**Comment by minimul** [comment]
I'll take a look soon.

**Comment by minimul** [comment]
I'll be taking a look this week.

**Comment by minimul** [comment]
Merged and released in 3.0.3.

**Pull request merged**